pkg/reconciler/instances/eventing/preaction/handletogglejsfilestorage.go

```diff
@@ -0,0 +1,205 @@
+)
+
+type handleToggleJSFileStorage struct {
+	kubeClientProvider
+}
+
+// newHandleEnableJSFileStorage
+func newHandleToggleJSFileStorage() *handleToggleJSFileStorage {
+	return &handleToggleJSFileStorage{
+		kubeClientProvider: defaultKubeClientProvider,
+	}
+}
+
+// Config holds the global configuration values.
+type Config struct {
+	Global Global
+}
+
+// Global configuration of Jetstream feature.
+type Global struct {
+	Jetstream Jetstream `yaml:"jetstream"`
+}
+
+// Jetstream specific values like enabled flag and storage type.
+type Jetstream struct {
+	Enabled string `yaml:"enabled"`
+	Storage string `yaml:"storage"`
+}
+
+func (r *handleToggleJSFileStorage) Execute(context *service.ActionContext, logger *zap.SugaredLogger) error {
+	// decorate logger
+	logger = logger.With(log.KeyStep, handleEnableJSFileStorageName)
+
+	kubeClient, err := r.kubeClientProvider(context, logger)
+	if err != nil {
+		return err
+	}
+
+	clientSet, err := kubeClient.Clientset()
+	if err != nil {
+		return err
+	}
+
+	statefulSet, err := clientSet.AppsV1().StatefulSets(namespace).Get(context.Context, statefulSetName, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.With(log.KeyReason, "Nats StatefulSet not found").Info("Step skipped")
+			return nil
+		}
+		return err
+	}
+
+	tracker, err := progress.NewProgressTracker(clientSet, logger, progress.Config{Interval: progressTrackerInterval, Timeout: progressTrackerTimeout})
+	if err != nil {
+		return err
+	}
+
+	comp := chart.NewComponentBuilder(context.Task.Version, eventingComponentName).
+		WithConfiguration(context.Task.Configuration).
+		WithNamespace(namespace).
+		Build()
+
+	chartValues, err := context.ChartProvider.Configuration(comp)
+	if err != nil {
+		return err
+	}
+
+	data, err := yaml.Marshal(chartValues)
+	if err != nil {
+		return err
+	}
+
+	values := &Config{}
+	if err := yaml.Unmarshal(data, &values); err != nil {
+		return err
+	}
+
+	// fetch Jetstream configuration
+	jetstreamEnabled := fmt.Sprint(values.Global.Jetstream.Enabled) == "true"
+	storageType := fmt.Sprint(values.Global.Jetstream.Storage)
+
+	// check if the Jetstream PVC is in place
+	volumeClaimTemplateFound := false
+	for _, volumeClaimTemplate := range statefulSet.Spec.VolumeClaimTemplates {
+		if strings.EqualFold(volumeClaimTemplate.Name, volumeClaimName) {
+			volumeClaimTemplateFound = true
+		}
+	}
+
+	//Scenario 1: migrate from NATS to NATS Jetstream with file storage
+	//as the statefulSet's spec is immutable, we cannot patch in order to add a PVC, hence statefulSet must be recreated
+	if jetstreamEnabled && strings.EqualFold(fileStorageType, fmt.Sprint(storageType)) {
+		if !volumeClaimTemplateFound {
+			if err := deleteNatsStatefulSet(context, clientSet, tracker, logger, statefulSet, false); err != nil {
+				return err
+			}
+		} else {
+			logger.With(log.KeyReason, "Nats StatefulSet already contains a PVC").Info("Step skipped")
+		}
+		return nil
+	}
```

Comment:
(nit) this nested ifs can be simplified:
```
if jetstreamEnabled && !volumeClaimTemplateFound && strings.EqualFold(fileStorageType, fmt.Sprint(storageType)) {
		logger.With(log.KeyReason, "Nats StatefulSet does not contains a PVC for JetStream File Storage").Info("Removing StatefulSet")
		if err := deleteNatsStatefulSet(context, clientSet, tracker, logger, statefulSet, false); err != nil {
			return err
		}
		return nil
}
```

Reply:
I know, but the reason I put the `if !volumeClaimTemplateFound {` one level below is because I wanted to output a meaningful `Step skipped` message. So it's done intentionally